pkg/zval/zval.go

```diff
@@ -45,13 +45,18 @@ func (i *Iter) Next() ([]byte, bool, error) {
 // AppendContainer appends to dst a zval container comprising the zvals in vals.
 func AppendContainer(dst []byte, vals [][]byte) []byte {
 	if vals == nil {
-		return AppendUvarint(dst, newTagUnset(true))
+		return AppendUvarint(dst, containerTagUnset)
 	}
 	var n int
 	for _, v := range vals {
-		n += sizeBytes(v)
+		//XXX this doesn't look like it could work right because we
+		// need to know whether the sub-zval is a container or a value,
+		// but in practice, the lengths are always the same because the
+		// variable length encoding of the size is not affected by the
+		// low bit of the tag encoding.
+		n += sizeOfValue(len(v))
```

Comment:
Though it's at all not clear from the doc comment (and that's on me), if `vals` contains any containers, you're going to get garbage from this function. Mentioning containers in this comment suggests otherwise.